### PR TITLE
[BEAM-1718] Handle Infinite Skew in SimpleDoFnRunner

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
Deprecate `getAllowedTimestampSkew`

### CODE REVIEW

**Comment by tgroh** [comment]
retest this please

**Comment by tgroh** [comment]
R: @jkff OR @kennknowles OR @bjchambers

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java [hunk 1]
```diff
@@ -575,8 +575,12 @@ public Instant timestamp() {
       return windowedValue.getWindows();
     }
 
+    @SuppressWarnings("deprecation") // Allowed Skew is deprecated for users, but must be respected
```
Comment (jkff) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java]: Add a unit test in SimpleDoFnRunnerTest?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java [hunk 2]
```diff
@@ -317,14 +318,18 @@ protected final void setupDelegateAggregators() {
...
+   * Returns the allowed timestamp skew duration, which is the maximum duration that timestamps can
+   * be shifted backward in {@link DoFn.Context#outputWithTimestamp}.
    *
-   * <p>The default value is {@code Duration.ZERO}, in which case
-   * timestamps can only be shifted forward to future.  For infinite
-   * skew, return {@code Duration.millis(Long.MAX_VALUE)}.
+   * <p>The default value is {@code Duration.ZERO}, in which case timestamps can only be shifted
+   * forward to future. For infinite skew, return {@code Duration.millis(Long.MAX_VALUE)}.
+   *
+   * @deprecated This method permits a {@link DoFn} to emit elements behind the watermark. These
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: Document what is the recommended replacement.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java [hunk 3]
```diff
@@ -92,7 +92,11 @@ private WithTimestamps(SerializableFunction<T, Instant> fn, Duration allowedTime
    * duration that timestamps can be shifted backwards from the timestamp of the input element.
    *
    * @see DoFn#getAllowedTimestampSkew()
+   * @deprecated This method permits a elements to be timestamped behind the watermark. These
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java]: Ditto.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 4]
```diff
@@ -1289,10 +1290,10 @@ public void testParDoShiftTimestamp() {
 
     PCollection<String> output =
         input
-        .apply(ParDo.of(new TestOutputTimestampDoFn()))
-        .apply(ParDo.of(new TestShiftTimestampDoFn(Duration.millis(1000),
+        .apply(ParDo.of(new TestOutputTimestampDoFn<Integer>()))
+        .apply(ParDo.of(new TestShiftTimestampDoFn<Integer>(Duration.millis(1000),
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Reformat this expression

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 5]
```diff
@@ -1306,11 +1307,15 @@ public void testParDoShiftTimestamp() {
   @Category(NeedsRunner.class)
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Hm, shouldn't all these tests be RunnableOnService?.. I'm still confused about how exactly it differs from NeedsRunner.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java [hunk 6]
```diff
@@ -1325,10 +1330,10 @@ public void testParDoShiftTimestampInvalid() {
   public void testParDoShiftTimestampInvalidZeroAllowed() {
 
     pipeline.apply(Create.of(Arrays.asList(3, 42, 6)))
-        .apply(ParDo.of(new TestOutputTimestampDoFn()))
-        .apply(ParDo.of(new TestShiftTimestampDoFn(Duration.ZERO,
+        .apply(ParDo.of(new TestOutputTimestampDoFn<Integer>()))
+        .apply(ParDo.of(new TestShiftTimestampDoFn<Integer>(Duration.ZERO,
```
Comment (jkff) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Reformat this

File: sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java [hunk 7]
```diff
@@ -174,6 +174,7 @@ public void parse(GridFSDBFile input, ParserCallback<String> callback)
         .setParser(TEXT_PARSER)
         .setCoder(StringUtf8Coder.of())
         .setConnectionConfiguration(ConnectionConfiguration.create())
+        .setSkew(Duration.ZERO)
```
Comment (jkff) [inline sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java]: Is this related?..

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: Currently there is none. https://issues.apache.org/jira/browse/BEAM-644 tracks this

File: runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java (hunk 1 above)
Comment (tgroh) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java]: Done

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java]: Done.

Also deprecated `withAllowedTimestampSkew()`, and added a "CAUTION" in the class documentation.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java (hunk 4 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Done.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java (hunk 5 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Mostly they're meant to be an interestingness barrier, excepting things like `TestStream`. I'll happily change these, as there doesn't appear to be any equivalent coverage.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java (hunk 6 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java]: Done.

File: sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java (hunk 7 above)
Comment (tgroh) [inline sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java]: Yes. `getAllowedTimestampSkew` is not permitted to return `null` but if unset `MongoDbGridFSIO` was doing so. Obtaining the value of `getMillis` from the returned skew was throwing a `NullPointerException` when running these tests.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java [hunk 8]
```diff
@@ -234,6 +239,97 @@ public void testOnTimerCalled() {
                 TimeDomain.EVENT_TIME)));
   }
 
+  @Test
+  public void testBackwardsInTimeNoSkew() {
```
Comment (jkff) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: Add short comments to all of these tests what they're testing.
E.g. this one is testing that with an allowed skew of 0, you can't shift backwards at all. The next one tests actual skew > allowed skew (non-zero), etc.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java [hunk 9]
```diff
@@ -234,6 +239,97 @@ public void testOnTimerCalled() {
...
+        String.format(
+            "the allowed skew (%s)",
+            PeriodFormat.getDefault().print(Duration.standardMinutes(10L).toPeriod())));
+    runner.processElement(
+        WindowedValue.timestampedValueInGlobalWindow(Duration.standardHours(1L), new Instant(0)));
+  }
+
+  @Test
+  public void testInfiniteSkew() {
+    SkewingDoFn fn = new SkewingDoFn(Duration.millis(Long.MAX_VALUE));
```
Comment (jkff) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: Is there a test for the case where actual skew < allowed skew and things should pass?

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java [hunk 10]
```diff
@@ -234,6 +239,97 @@ public void testOnTimerCalled() {
...
+
+    runner.startBundle();
+    runner.processElement(
+        WindowedValue.timestampedValueInGlobalWindow(Duration.millis(1L), new Instant(0)));
+    runner.processElement(
+        WindowedValue.timestampedValueInGlobalWindow(
+            Duration.millis(1L), BoundedWindow.TIMESTAMP_MIN_VALUE.plus(Duration.millis(1))));
+    runner.processElement(
+        WindowedValue.timestampedValueInGlobalWindow(
+            // The maximum distance
```
Comment (jkff) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: Please expand this comment, I'm not following.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java [hunk 11]
```diff
@@ -327,7 +327,7 @@ protected final void setupDelegateAggregators() {
    * @deprecated This method permits a {@link DoFn} to emit elements behind the watermark. These
    *     elements are considered late, and if behind the
    *     {@link Window#withAllowedLateness(Duration) allowed lateness} of a downstream
-   *     {@link PCollection} may be silently dropped.
+   *     {@link PCollection} may be silently dropped. See BEAM-644 for details on a replacement.
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: Give full http:// link?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java [hunk 12]
```diff
@@ -45,10 +45,15 @@
...
-   * {@link #withAllowedTimestampSkew(Duration)} to update the allowed skew.
+   * element must not be before the input element's timestamp minus the value of {@link
+   * #getAllowedTimestampSkew()}. If an output timestamp is before this time, the transform will
+   * throw an {@link IllegalArgumentException} when executed. Use {@link
+   * #withAllowedTimestampSkew(Duration)} to update the allowed skew.
+   *
+   * <p>CAUTION: Use of {@link #withAllowedTimestampSkew(Duration)} permits elements to be emitted
+   * behind the watermark. These elements are considered late, and if behind the {@link
+   * Window#withAllowedLateness(Duration) allowed lateness} of a downstream {@link PCollection} may
+   * be silently dropped. See BEAM-644 for details on a replacement.
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java]: Ditto

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java [hunk 13]
```diff
@@ -82,7 +87,12 @@ private WithTimestamps(SerializableFunction<T, Instant> fn, Duration allowedTime
    *
    * <p>The default value is {@code Duration.ZERO}, allowing timestamps to only be shifted into the
    * future. For infinite skew, use {@code new Duration(Long.MAX_VALUE)}.
+   * @deprecated This method permits a to elements to be emitted behind the watermark. These
+   *     elements are considered late, and if behind the
+   *     {@link Window#withAllowedLateness(Duration) allowed lateness} of a downstream
+   *     {@link PCollection} may be silently dropped. See BEAM-644 for details on a replacement.
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java]: Ditto

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java [hunk 14]
```diff
@@ -92,9 +102,10 @@ private WithTimestamps(SerializableFunction<T, Instant> fn, Duration allowedTime
    * duration that timestamps can be shifted backwards from the timestamp of the input element.
    *
    * @see DoFn#getAllowedTimestampSkew()
-   * @deprecated This method permits a elements to be timestamped behind the watermark. These
-   *     elements are considered late, and if behind the {@link Window#withAllowedLateness(Duration)
-   *     allowed lateness} of a downstream {@link PCollection} may be silently dropped.
+   * @deprecated This method permits a to elements to be emitted behind the watermark. These
+   *     elements are considered late, and if behind the
+   *     {@link Window#withAllowedLateness(Duration) allowed lateness} of a downstream
+   *     {@link PCollection} may be silently dropped. See BEAM-644 for details on a replacement.
```
Comment (jkff) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/WithTimestamps.java]: Ditto

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java (hunk 8 above)
Comment (tgroh) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: Done.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java (hunk 9 above)
Comment (tgroh) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: Yes, it's included in `testSkew`, before the exception.

File: runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java (hunk 10 above)
Comment (tgroh) [inline runners/core-java/src/test/java/org/apache/beam/runners/core/SimpleDoFnRunnerTest.java]: Updated.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java (hunk 11 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java]: Done
